### ISSUE TITLE
Stopped working after upgrade to tide-20170210.1932

### ISSUE THREAD

**Issue by edoloughlin** [ananthakumaran/tide#114]
After upgrading to 20170210.1932, I'm constantly seeing:
Error running timer ‘tide--hl-identifiers-function’: (error "Server does not exist. Run M-x tide-restart-server to start it again")
If I restart the server then I can see the tsserver process start and then immediately exit. I can't see an error message.
I'm on Windows 7, GNU Emacs 25.1.1 (x86_64-w64-mingw32) of 2016-09-17

**Comment by ananthakumaran** [comment]
Can you post the contents of the log file. Look at the readme file for how to enable the log.

**Comment by ksjogo** [comment]
It's not liking the newlines (anymore?) and is triggering command execution on each, so I removed them in the PR above.

**Comment by edoloughlin** [comment]
I can't seem to enable the log. No log is produced if I set TSS_LOG according to the readme. If I don't specify the -file arg then no log is produced in the default location.

**Comment by ksjogo** [comment]
Strange. I just did the eval-expression for it and the log was created.

**Comment by edoloughlin** [comment]
@ksjogo Unfortunately, I'm on Windows 7. I'm not sure how the path is interpreted, but I've tried every variation I can think of and there's still no log. I applied your patch locally but it hasn't fixed the issue.

**Comment by edoloughlin** [comment]
I've tried this with emacs started from Explorer and Cygwin. I thought the issue might be my log path but I now suspect that tsserver is exiting before it reads the config.

**Comment by edoloughlin** [comment]
For what it's worth, this is what I see in *Messages*. The 'Invalid argument' message only appears once, so I missed it before:
```
(src) Starting tsserver...
(src) tsserver server started successfully.
Error while checking syntax automatically: (file-error "Writing to process" "Invalid argument" #<process tsserver>)
(src) tsserver exits: exited abnormally with code 1.
Error running timer ‘tide--hl-identifiers-function’: (error "Server does not exist. Run M-x tide-restart-server to start it again")
```

**Comment by mzgnr** [comment]
@ananthakumaran im on os x and experiencing same issue :(

**Comment by ananthakumaran** [comment]
@mzgnr can you post emacs version, tsserver-version, and the contents of the log file? I am not sure whether it's a regression in emacs/tide/tsserver.

**Comment by kraf** [comment]
I have this problem too.

- OSX El Capitan
- Emacs 25.1.1

Logs:

```
Info 0    Binding...
Info 1    request: {"command":"open","seq":"14","arguments":{"file":"/***/app/server/index.ts","scriptKindName":"TS"}}
Info 2    Search path: /***/app/server
Info 3    Config file name: /***/app/server/tsconfig.json
Info 4    Add recursive watcher for: /***/app/server
Info 5    Add type root watcher for: /***/node_modules/@types
Info 6    Opened configuration file /***/app/server/tsconfig.json
Info 7    Project '/***/app/server/tsconfig.json' (Configured) 0
/***/.emacs.d/elpa/tide-20170210.1932/tsserver/lib.es6.d.ts
/***/node_modules/@types/node/index.d.ts
/***/node_modules/@types/keygrip/index.d.ts
/***/node_modules/@types/cookies/index.d.ts
/***/node_modules/@types/koa/index.d.ts
/***/node_modules/@types/koa-router/index.d.ts
/***/app/server/index.ts
/***/node_modules/@types/dom4/index.d.ts
/***/node_modules/@types/es6-promise/index.d.ts
/***/node_modules/@types/koa-static/index.d.ts
/***/node_modules/@types/react/index.d.ts
/***/node_modules/@types/react-addons-transition-group/index.d.ts
/***/node_modules/@types/react-addons-css-transition-group/index.d.ts
/***/node_modules/@types/react-dom/index.d.ts
/***/node_modules/redux/index.d.ts
/***/node_modules/@types/react-redux/index.d.ts
/***/node_modules/@types/redux-immutable/index.d.ts
/***/node_modules/@types/tether/index.d.ts

-----------------------------------------------
Open files:
/***/app/server/index.ts
Info 8    event: {"seq":0,"type":"event","event":"configFileDiag","body":{"triggerFile":"/***/app/server/index.ts","configFile":"/***/app/server/tsconfig.json","diagnostics":[]}}
Perf 9    14::open: async elapsed time (in milliseconds) 712.9128
Info 10   request: {"command":"configure","seq":"15","arguments":{"hostInfo":"GNU Emacs 25.1.1 (x86_64-apple-darwin13.4.0, NS appkit-1265.21 Version 10.9.5 (Build 13F1911))\n of 2016-09-21","file":"/***/app/server/index.ts","formatOptions":{"tabSize":8,"indentSize":4}}}
Info 11   Host configuration update for file /***/app/server/index.ts
Info 12   response: {"seq":0,"type":"response","command":"configure","request_seq":"15","success":true}
Perf 13   15::configure: async elapsed time (in milliseconds) 1.1424
```

**Comment by ananthakumaran** [comment]
@kraf can you post the emacs error as well? I don't see anything wrong with the log

**Comment by kraf** [comment]
Sure, thanks for the quick reply:

```
(server) Starting tsserver...
(server) tsserver server started successfully.
Error in post-command-hook (tide--hl-identifiers-post-command-hook): (void-function incf)
Error running timer ‘tide--hl-identifiers-function’: (void-function incf) [18 times]
Type "q" in help window to restore its previous buffer.
Error running timer ‘tide--hl-identifiers-function’: (void-function incf) [9 times]
Quit [2 times]
Mark set
("TSS_LOG=-level verbose -file /tmp/tss.log")
Saving file /Users/filip/src/mcms/docker-compose.yml...
Wrote /Users/filip/src/mcms/docker-compose.yml
Quit [2 times]
Error running timer ‘tide--hl-identifiers-function’: (void-function incf)
(server) tsserver exits: killed: 9.
(server) Starting tsserver...
(server) tsserver server started successfully.
Error running timer ‘tide--hl-identifiers-function’: (void-function incf) [4 times]
```

**Comment by ananthakumaran** [comment]
@kraf can you try master and see if it fixes the problem?

**Comment by kraf** [comment]
Hi sorry it took a while, it had to clone TypeScript. Master works 👍

**Comment by edoloughlin** [comment]
I just installed tide-20170220.133 and it's all good now. It's also creating the log file. Thanks!

**Comment by valerian** [comment]
I just installed `tide-20170412.541` via elpa, on Windows 8.1, using Emacs 25 within the terminal on Cygwin x64, and I'm experiencing the exact same issue as @edoloughlin

I used to experience it too about less than a year ago, but had given up on it and ran a debian virtual machine instead.

However I just decided to try again, and stumbled upon this issue, so it seems to have been fixed at some point, but today for me it's not working.

I tried also with the latest version from master, then with a recompiled tsserver, and also with 35686d7 from Feb 20, which should be the one that worked for @edoloughlin, but still no luck.

Any tips would be greatly appreciated :)

**Comment by josteink** [comment]
Maybe setting edebug-on-error to t might help highlight what's actually causing the error?

**Comment by valerian** [comment]
@josteink I'm really not familiar with edebug and lisp debugging in general, could you please help me a bit on how to do that?

**Comment by ananthakumaran** [comment]
@valerian looks like tide is not able to start the tsserver. You can try to start tsserver from command line and check if it's working properly. You can get the path of tsserver that comes with tide by running `M-x eval-expression` and type `(expand-file-name "tsserver.js" tide-tsserver-directory)` in minibuffer. This would give you the path. Then run `node path` in command line.

**Comment by josteink** [comment]
Just calling in to say that I'm using tide almost daily, on Windows, and I currently don't have any issues. I recall having similar issues in the past.

I may be wrong, but I'm pretty sure it was a local configuration issue. For the record: I have now customized `tide-tsserver-executable` to point to a a globally installed TypeScript folder instead of what ships with `tide`.

You could always try that?

**Comment by josteink** [comment]
Speaking of which, new TypeScript and new tsserver is out. We should update :)

@ananthakumaran:  Do you just delete the existing `tsserver`-folder and replace it with whatever is in `$HOME/.npm/node_modules/typescript/lib/tsserver`? If so, I can probably get the job done too.

**Comment by ananthakumaran** [comment]
@josteink There is a Makefile in root folder to automate some part of it. You have to run `make setup` first time.

1) go to typescript repo and checkout the correct tag.
2) go to tide repo and run `make update-tsserver`. This will build & copy required files. we need a single js file and all the definition files.
3) do some testing to make sure everything works.
4) I keep the version of tsserver and tide in sync, next step would be to update the version number in tide.el & create a new tag and push.

**Comment by valerian** [comment]
One thing I realized is that I had to create a few symbolic links for node and tsserver to be able to properly translate cygwin paths to windows paths (for me `C:\home` to `C:\cygwin64\home`, and `C:\cygdrive\c` to `C:\`), but it didn't solve the problem.

@ananthakumaran thanks, I already tried recompiling tsserver with your makefile yesterday and it didn't help. When trying from command line to start either the one provided with tide (both Elpa or Github versions), or the one generated from `make update-tsserver`, I get this error:
```module.js:472
    throw err;
    ^

Error: Cannot find module 'C:\cygwin64\home\valerian\.emacs.d\elpa\tide-20170412.541\tsserver\typingsInstaller.js'
    at Function.Module._resolveFilename (module.js:470:15)
    at Function.Module._load (module.js:418:25)
    at Module.runMain (module.js:605:10)
    at run (bootstrap_node.js:423:7)
    at startup (bootstrap_node.js:147:9)
    at bootstrap_node.js:538:3```

Although this does not seem to be a fatal error, since the tsserver prompt is still there are seems to work.

The tsserver from the globally installed typescript npm module works, which is located for me at `/cygdrivee/c/Users/Valerian/AppData/Roaming/npm/tsserver`. I had tried using it by doing `(setq tide-tsserver-executable "tsserver")` or `(setq tide-tsserver-executable "/cygdrivee/c/Users/Valerian/AppData/Roaming/npm/tsserver")`, but then I realized just now that I was obviously supposed to give the path to `tsserver.ts`!

**UPDATE:**

So right now, while writing this answer, I just solved my problem :)
This is what I needed to do:

`(setq tide-tsserver-executable "/cygdrive/c/Users/Valérian/AppData/Roaming/npm/node_modules/typescript/lib/tsserver.js")`

**Comment by josteink** [comment]
Yeah. Mixed Cygwin / native setups are a nightmare. 

Glad you got things worked out.

**Comment by josteink** [comment]
@ananthakumaran And there tsserver should be up to date. Had to add some new files. Feel free to review the update: https://github.com/ananthakumaran/tide/commit/26f6645c1de8c94d8c41755580abe3cb39ee749f

**Comment by ananthakumaran** [comment]
Thanks, works well.

**Comment by rtm** [comment]
Experiencing a similar problem, and then upgraded to the May 15 2018, but the problem persisted. I'm on WSL. Everything was working fine and then inexplicably stopped working:

    Error from syntax checker typescript-tide: UNKNOWN!
    Server does not exist. Run M-x tide-restart-server to start it again

    (src-b370818942) tsserver exits: exited abnormally with code 1.

The tsserver used by tide runs fine from the command line when invoked by node.

The command `(setq tide-tsserver-process-environment '("TSS_LOG=-level verbose -file /tmp/tss.log"))` does not seem to create a log.

What's the recommended way to go about debugging this?

emacs version is 24.5.1.

Here's something from `*Message*`:

```
Couldn't locate project root folder with a tsconfig.json or jsconfig.json file. Using '/home/rtm/' as project root.
(rtm-85a69add72) Starting tsserver...
(rtm-85a69add72) tsserver server started successfully.
File mode specification error: (error "Process tsserver not running")
File local-variables error: (error "Process tsserver not running")
Error while checking syntax automatically: (error "Process tsserver not running")
Wrote /home/rtm/.emacs.d/.emacs.desktop.lock
Desktop: 1 frame, 3 buffers restored.
For information about GNU Emacs and the GNU system, type C-h C-a.
(rtm-85a69add72) tsserver exits: exited abnormally with code 1.
```

**Comment by ambirdsall** [comment]
I've got the same issue reported above, using `tide-20180810` on macOS (High Sierra): `tsserver` is immediately exiting with code 1 and never creating a log file. I set `tide-tsserver-process-environment` to `"TSS_LOG=-level verbose -file /tmp/tss.log"` and `tide-tsserver-executable` to my work project's `node_modules/typescript/bin/tsserver`. Manually calling that same executable on the command line doesn't error.

I'm not sure what else I can do to debug this; any help would be greatly appreciated.
